JPEGImageCompression/image_compression.py

```diff
@@ -357,19 +364,242 @@ def update(mob, alpha):
+        jpeg_decoder = self.module("JPEG Decoder")
+
+        data_flow_decode = (
+            VGroup(compressed_data.copy(), jpeg_decoder, final_image)
+            .arrange(RIGHT, buff=1)
+            .shift(DOWN * 2)
+        )
+        arr3 = Arrow(data_flow_decode[0].get_right(), jpeg_decoder.get_left())
+        arr4 = Arrow(jpeg_decoder.get_right(), final_image.get_left())
+        data_flow_decode.add(arr3, arr4)
+
+        # animations
+        self.play(LaggedStartMap(Write, input_image, lag_ratio=0.1), run_time=2)
+
+        self.play(
+            Write(jpeg_encoder),
+            Write(arr1),
+        )
+        self.play(
+            Write(compressed_data),
+            Write(arr2),
+        )
+
+        self.wait()
+
+        self.play(data_flow_encode.animate.shift(UP * 2))
+
+        self.wait()
+
+        self.play(LaggedStartMap(Write, data_flow_decode, lag_ratio=0.1), run_time=1)
+
+        self.wait(4)
+
+    def zoom_jpeg_encoder(self):
+        jpeg_encoder = self.module("JPEG Encoder")
+        self.play(Write(jpeg_encoder.move_to(ORIGIN)))
+        self.wait()
+        self.play(
+            ScaleInPlace(jpeg_encoder, 3),
+            FadeOut(jpeg_encoder[1]),
+        )
+
+        forward_dct = self.module(
+            "Forward DCT", fill_color="#7F7F2D", stroke_color=REDUCIBLE_YELLOW
+        )
+        forward_dct_icon = ImageMobject("dct.png").scale(0.4)
+        forward_dct_module = Group(forward_dct, forward_dct_icon).arrange(DOWN, buff=1)
+
+        quantization = self.module(
+            "Quantization", fill_color="#7F7F2D", stroke_color=REDUCIBLE_YELLOW
+        )
+        quantization_icon = ImageMobject("quantization.png").scale(0.4)
+        quantization_module = Group(quantization, quantization_icon).arrange(
+            DOWN, buff=1
+        )
+
+        lossless_comp = self.module(
+            "Lossless \\\\ compression",
+            fill_color="#7F7F2D",
+            stroke_color=REDUCIBLE_YELLOW,
+        )
+        lossless_icon = ImageMobject("lossless.png").scale(0.4)
+        lossless_module = Group(lossless_comp, lossless_icon).arrange(DOWN, buff=1)
+
+        modules_g = (
+            Group(forward_dct_module, quantization_module, lossless_module)
+            .arrange(RIGHT)
+            .scale_to_fit_width(jpeg_encoder.width - 1)
+        )
+
+        self.play(LaggedStartMap(FadeIn, modules_g), run_time=2)
+
+        self.wait(3)
+
+    def intro_2(self):
+
+        img_og = self.make_nested_squares(8, 2).scale_to_fit_width(4)
+
+        img_sm = self.make_nested_squares(4, 1).scale_to_fit_width(4)
+
+        self.play(FadeIn(img_og), run_time=2)
+
+        anims = []
+        for i in range(len(img_og.submobjects)):
+            anims.append(Transform(img_og[i], img_sm[i]))
+
+        self.play(LaggedStart(*anims, lag_ratio=0.05), run_time=10)
+
+        self.wait(3)
+
+    # definition of util functions
+    def make_nested_squares(self, total_side=8, groups_side=2):
```

Comment:
I think this is meant to show a smaller version of the image, but to me, it's not super clear. Looks more just like a lower resolution image, and I'm not sure the viewer will make that connection. Also, JPEG is not actually lowering the resolution, so might be confusing.